library_registry/admin/controller.py

```diff
@@ -18,33 +23,65 @@
     Validation,
 )
 
+
 class ViewController(BaseController):
     def __call__(self):
-        username = session.get('username', '')
+        """View controller for setting Flask Session or Flask JWT to verify user
+
+        Returns:
+            Response: rednered template string with user identity
+        """
+        if verify_jwt_in_request(optional=True):
+            username = get_jwt_identity()
+        else:
+            username = session.get('username', '')
+
         response = Response(render_template_string(
             admin_template,
             username=username
         ))
         return response
 
+
 class AdminController(BaseController):
 
     def __init__(self, app, emailer_class=Emailer):
         super(AdminController, self).__init__(app)
         self.emailer = emailer_class
-    
-    def log_in(self):
+
+    def log_in(self, jwt_cookie_boolean=False):
+        """End point for login with requesting flask sessions or JWT tokens
+
+        Returns:
+            err: Invalid credentials
+            flask session: flask session and redirect
+            jwt tokens: JWT tokens as Set-Cookie headers and redirect
+        """
         username = request.form.get("username")
         password = request.form.get("password")
-        if Admin.authenticate(self._db, username, password):
+        if not Admin.authenticate(self._db, username, password):
+            return INVALID_CREDENTIALS
+        if not jwt_cookie_boolean:
             session["username"] = username
             return redirect(url_for('admin.admin_view'))
-        else:
-            return INVALID_CREDENTIALS
+        access_token = create_access_token(identity=username)
+        response = make_response(redirect(url_for('admin.admin_view')), 302)
```

Comment:
still have to raise a question here! when requesting a JWT I still don't think the redirect makes sense. This will redirect to the `/admin` route that serves the locally deployed version of the admin interface. If the request _doesn't_ come from that instance, it breaks right?

I think the right thing to do would be to return a `200` with the token in the body of the response.

Reply:
I admit that it took me a bit and probably too long to wrap my head around this because I was thinking in terms of monolithic software vs a decoupled architecture. I think the example that made me finally understand why **this is the right approach** is: we're building and maintaining an API. In theory, multiple clients could be calling our API. It makes no sense to send back to everyone a redirect to a URL that is specific to only one particular client. Thanks very much, Mike!!!!